x/epoch/types/hooks.go

```diff
@@ -0,0 +1,39 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type EpochHooks interface {
+	// the first block whose timestamp is after the duration is counted as the end of the epoch
+	AfterEpochEnd(ctx sdk.Context, epoch Epoch)
+	// new epoch is next block of epoch end block
+	BeforeEpochStart(ctx sdk.Context, epoch Epoch)
+}
+
+var _ EpochHooks = MultiEpochHooks{}
+
+// combine multiple gamm hooks, all hook functions are run in array sequence.
+type MultiEpochHooks []EpochHooks
+
+func NewMultiEpochHooks(hooks ...EpochHooks) MultiEpochHooks {
+	return hooks
+}
+
+// AfterEpochEnd is called when epoch is going to be ended, epochNumber is the number of epoch that is ending.
+func (h MultiEpochHooks) AfterEpochEnd(ctx sdk.Context, epoch Epoch) {
+	for i := range h {
+		cacheCtx, write := ctx.CacheContext()
+		h[i].AfterEpochEnd(cacheCtx, epoch)
+		write()
```

Comment:
cache would not be flushed only if `AfterEpochEnd` panicks, which would halt the chain anyway before `Commit`, so we might not need CacheContext here?

Reply:
good point, since we just panic i think it's fine to use original ctx